code/modules/power/prototype_shield_emitter.dm

```diff
@@ -0,0 +1,208 @@
+	anchored = TRUE
+
+/obj/machinery/power/proto_sh_emitter/Destroy()
+	if(SSticker.IsRoundInProgress())
+		var/turf/T = get_turf(src)
+		message_admins("Prototype Shield Emitter deleted at [ADMIN_VERBOSEJMP(T)]")
+		log_game("Prototype Shield Emitter deleted at [AREACOORD(T)]")
+	for(var/H in signs)
+		qdel(H)
+	return ..()
+
+/obj/machinery/power/proto_sh_emitter/update_icon_state()
+	if(is_on == TRUE)
+		icon_state = "proto_sh_emitter_on"
+	else
+		icon_state = "proto_sh_emitter"
+
+/obj/machinery/power/proto_sh_emitter/proc/can_be_rotated(mob/user,rotation_type)
+	if(anchored)
+		to_chat(user, "<span class='warning'>It is fastened to the floor!</span>")
+		return FALSE
+	return TRUE
+
+/obj/machinery/power/proto_sh_emitter/wrench_act(mob/living/user, obj/item/I)
+	if(is_on == TRUE)
+		to_chat(user, "<span class='warning'>You have to turn the [src] off first!</span>")
+		return TRUE
+	anchored = !anchored
+	if(anchored)
+		to_chat(user, "<span class='warning'>You bolt the [src] to the floor!</span>")
+	else
+		to_chat(user, "<span class='warning'>You unbolt the [src] from the floor!</span>")
+	return TRUE
+
+/obj/machinery/power/proto_sh_emitter/attackby(obj/item/I, mob/user, params)
+	if(I.GetID())
+		if(allowed(user))
+			if(is_on)
+				locked = !locked
+				to_chat(user, "<span class='notice'>You [locked ? "lock" : "unlock"] the controls.</span>")
+			else
+				to_chat(user, "<span class='warning'>The controls can only be locked when \the [src] is online!</span>")
+		else
+			to_chat(user, "<span class='danger'>Access denied.</span>")
+		return
+
+/obj/machinery/power/proto_sh_emitter/crowbar_act(mob/living/user, obj/item/I)
+	default_deconstruction_crowbar(I)
+	return TRUE
+
+/obj/machinery/power/proto_sh_emitter/screwdriver_act(mob/living/user, obj/item/I)
+	if(..())
+		return TRUE
+	default_deconstruction_screwdriver(user, "proto_sh_emitter-open", "proto_sh_emitter", I)
+	return TRUE
+
+/obj/machinery/power/proto_sh_emitter/process()
+	var/area/a = get_area(src)
+	var/turf/Turf = get_turf(src)
+	if(a.power_equip == FALSE && is_on == TRUE)
+		is_on = FALSE
+		update_icon_state()
+		message_admins("[src] turned off at [ADMIN_VERBOSEJMP(Turf)]")
+		log_game("[src] turned off at [AREACOORD(Turf)]")
+
+/obj/machinery/power/proto_sh_emitter/interact(mob/user)
+	var/area/a = get_area(src)
+	add_fingerprint(user)
+	if(!anchored)
+		to_chat(user, "<span class='warning'>You need to anchor the [src] first!</span>")
+		return
+	if(a.power_equip == FALSE)
+		to_chat(user, "<span class='warning'>There is no power in this area!!</span>")
+		return
+	if(locked)
+		to_chat(user, "<span class='warning'>The controls are locked!</span>")
+		return
+	if(is_on == TRUE)
+		remove_barrier(user)
+		return
+	check_map_borders(2,5,2,5)
+	if(!borders)
+		to_chat(user, "<span class='warning'>The motors whir and fail!</span>")
+		return
+	build_barrier(1,2,1,4,2,1,2,5,user)
+
+/obj/machinery/power/proto_sh_emitter/small/interact(mob/user)
```

Comment:
This is some copypaste, if they differ just with the actual barrier parameters wrap that in a proc

Reply:
what should i wrap in a proc? the interact stuff or the build_barrier() stuff?

Reply:
The common checks in one function, the barrier creation in another called from that and overriden for subtypes.
Or you can just put these arguments in machine variables and change them in the subtype